// apollo-engine/lib/__tests__/engine-restify.test.js
"use strict";
// We put the restify tests in their own file because importing restify breaks
// instances of express in the same process:
// https://github.com/restify/node-restify/issues/1540
Object.defineProperty(exports, "__esModule", { value: true });
const apollo_server_restify_1 = require("apollo-server-restify");
const restify = require("restify");
const schema_1 = require("./schema");
const engine_common_1 = require("./engine-common");
engine_common_1.runSuitesForHttpServerFramework('restify', {
    createApp() {
        const server = restify.createServer({ name: 'test server' });
        const graphQLOptions = {
            schema: schema_1.schema,
            rootValue: schema_1.rootValue,
            tracing: true,
            cacheControl: true,
        };
        server.use(restify.plugins.bodyParser());
        server.use(restify.plugins.queryParser());
        server.post('/graphql', apollo_server_restify_1.graphqlRestify(graphQLOptions));
        server.get('/graphql', apollo_server_restify_1.graphqlRestify(graphQLOptions));
        server.get('/graphql/ping', (req, res, next) => {
            res.send({ pong: true });
            next();
        });
        return server;
    },
    serverForApp(app) {
        return app.server;
    },
    appParameter: 'restifyServer',
});
