// apollo-engine/lib/__tests__/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const request = require("request");
exports.schema = graphql_1.buildSchema(`
  type Query {
    hello: String @cacheControl(maxAge: 30)
    errorTrigger: String
  }
`);
exports.rootValue = {
    hello: () => {
        return 'Hello World';
    },
    errorTrigger: () => {
        throw new Error('Kaboom');
    },
};
function verifyEndpointSuccess(url, hasTracing) {
    return new Promise(resolve => {
        request.post({
            url,
            json: true,
            body: { query: '{ hello }' },
        }, (err, response, body) => {
            expect(err).toBe(null);
            expect(body['data']['hello']).toBe('Hello World');
            if (hasTracing) {
                expect(body['extensions'] && body['extensions']['tracing']).toBeDefined();
            }
            else {
                expect(body['extensions'] && body['extensions']['tracing']).toBeUndefined();
            }
            resolve(body);
        });
    });
}
exports.verifyEndpointSuccess = verifyEndpointSuccess;
function verifyEndpointBatch(url, hasTracing) {
    return new Promise(resolve => {
        request.post({
            url,
            json: true,
            body: [{ query: '{ hello }' }, { query: '{ hello }' }],
        }, (err, response, bodies) => {
            expect(err).toBe(null);
            expect(bodies.length).toBe(2);
            bodies.forEach((body) => {
                expect(body['data']['hello']).toBe('Hello World');
                if (hasTracing) {
                    expect(body['extensions'] && body['extensions']['tracing']).toBeDefined();
                }
                else {
                    expect(body['extensions'] && body['extensions']['tracing']).toBeUndefined();
                }
            });
            resolve();
        });
    });
}
exports.verifyEndpointBatch = verifyEndpointBatch;
function verifyEndpointFailure(url) {
    return new Promise(resolve => {
        request.post({
            url,
            json: true,
            body: { query: '{ validButDoesNotComplyToSchema }' },
        }, (err, response, body) => {
            if (response.statusCode === 200) {
                // Proxy responds with an error-ed 200:
                expect(response.body['errors'][0]['message']).toBe('Cannot query field "validButDoesNotComplyToSchema" on type "Query".');
            }
            else {
                // Express responds with a 400
                expect(response.statusCode).toBe(400);
            }
            resolve();
        });
    });
}
exports.verifyEndpointFailure = verifyEndpointFailure;
function verifyEndpointError(url) {
    return new Promise(resolve => {
        request.post({
            url,
            json: true,
            body: { query: '{ errorTrigger }' },
        }, (err, response, body) => {
            expect(err).toBe(null);
            expect(response.statusCode).toBe(200);
            expect(body['errors'][0]['message']).toBe('Kaboom');
            resolve();
        });
    });
}
exports.verifyEndpointError = verifyEndpointError;
function verifyEndpointGet(url, hasTracing) {
    return new Promise(resolve => {
        let query = '{ hello }';
        request.get({
            url: `${url}?query=${encodeURIComponent(query)}`,
            json: true,
        }, (err, response, body) => {
            expect(err).toBe(null);
            expect(response.statusCode).toBe(200);
            expect(body['data']['hello']).toBe('Hello World');
            if (hasTracing) {
                expect(body['extensions'] && body['extensions']['tracing']).toBeDefined();
            }
            else {
                expect(body['extensions'] && body['extensions']['tracing']).toBeUndefined();
            }
            resolve();
        });
    });
}
exports.verifyEndpointGet = verifyEndpointGet;
